crates/bevy_sprite/src/frustum_culling.rs

```diff
@@ -0,0 +1,108 @@
+use std::time::Instant;
+
+use bevy_ecs::{Commands, Entity, Query, Res, With};
+use bevy_math::Vec2;
+use bevy_render::{
+    camera::Camera,
+    prelude::{Draw, Visible},
+    render_graph::base::MainPass,
+};
+use bevy_transform::components::Transform;
+use bevy_window::Windows;
+
+use crate::Sprite;
+
+struct Rect {
+    position: Vec2,
+    size: Vec2,
+}
+
+impl Rect {
+    #[inline]
+    pub fn contains(&self, point: Vec2) -> bool {
+        point.x >= self.position.x
+            && point.y >= self.position.y
+            && point.x < self.end().x
+            && point.y < self.end().y
+    }
+
+    #[inline]
+    pub fn end(&self) -> Vec2 {
+        self.position + self.size
+    }
+}
+
+pub fn frustum_culling(
+    commands: &mut Commands,
+    windows: Res<Windows>,
+    visible: Query<&Visible>,
+    cameras: Query<&Transform, With<Camera>>,
+    mut drawables: Query<(Entity, &Transform, &Sprite)>,
+) {
+    let start = Instant::now();
+
+    let window = windows.get_primary().unwrap();
+    let window_size = Vec2::new(window.width(), window.height());
+
+    for camera_transform in cameras.iter() {
+        let camera_position = Vec2::new(
+            camera_transform.translation.x,
+            camera_transform.translation.y,
+        );
+
+        let rect = Rect {
+            position: camera_position - window_size / Vec2::new(2.0, 2.0),
+            size: window_size,
+        };
+
+        for (entity, drawable_transform, sprite) in drawables.iter_mut() {
+            let half_width = sprite.size.x / 2.0;
+            let half_height = sprite.size.y / 2.0;
+
+            let corner1 = Vec2::new(
+                drawable_transform.translation.x - half_width,
+                drawable_transform.translation.y - half_height,
+            );
+
+            let corner2 = Vec2::new(
+                drawable_transform.translation.x + half_width,
+                drawable_transform.translation.y - half_height,
+            );
+
+            let corner3 = Vec2::new(
+                drawable_transform.translation.x - half_width,
+                drawable_transform.translation.y + half_height,
+            );
+
+            let corner4 = Vec2::new(
+                drawable_transform.translation.x + half_width,
+                drawable_transform.translation.y + half_height,
+            );
+
+            if rect.contains(corner1)
+                || rect.contains(corner2)
+                || rect.contains(corner3)
+                || rect.contains(corner4)
```

Comment:
If the sprite is larger than the window, it might accidentally get culled here (imagine say a large background image). Checking whether the window rect and the sprites rect intersect would be preferrable in my opinion and would guard against this edge case.

Reply:
This is a very good point. I'll change it.

Reply:
Done